clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -1121,13 +1140,25 @@ public long maybeUpdate(long now) {
 
             // Beware that the behavior of this method and the computation of timeouts for poll() are
             // highly dependent on the behavior of leastLoadedNode.
-            Node node = leastLoadedNode(now);
-            if (node == null) {
+            LeastLoadedNode leastLoadedNode = leastLoadedNode(now);
+
+            // Rebootstrap if needed and configured.
+            if (!leastLoadedNode.hasNodeAvailableOrConnectionReady()
+                    && metadataRecoveryStrategy == MetadataRecoveryStrategy.REBOOTSTRAP) {
+                for (final Node oldNode : metadata.fetch().nodes()) {
+                    NetworkClient.this.close(oldNode.idString());
```

Comment:
Is there a chance that there may be connections to old nodes? I think we want to rebootstrap only if every node is in backoff state. We don't close nodes in the Admin client equivalent below, which is confusing, so it may be better to remove this unless it is required. Checking `hasNodeAvailableOrConnectionReady` adds to the confusion since it suggests there may be nodes in some non-ready state that need to be closed. Should we change `leastLoadedNode()` method to return `hasConnectedOrConnecting` nodes instead of `atLeastOneConnectionReady` to make this more obvious?

Reply:
Oh that's right: if `!leastLoadedNode.hasNodeAvailableOrConnectionReady()`, there shouldn't be open connections. I removed this closing.